### PR TITLE
[jit] stop parsing the block after seeing exit statements

### LINKED ISSUE #44864: [JIT] treat `assert not torch.jit.is_scripting()` statically
## 🚀 Feature
JIT should treat `assert not torch.jit.is_scripting()` statically (like how it treats `if is_scripting()` and ignore what's after the assertion in a basicblock.

## Motivation

Having more ways to let torchscript ignore code sections will in general make it easier to migrate code to partially support torchscript.

## Pitch
Currently, this is scriptable:
```
class A(nn.Module):
    def __init__(self):
        super().__init__()
        self.k = 1.0
    def forward(self, x):
        if self.training and not torch.jit.is_scripting():
            return x * self.k if self.k != "test" else x   # not supported due to lack of union type
        else:
            return x + 1

a = A()
a = torch.jit.script(a)
```
but this fails:
```
class A(nn.Module):
    def __init__(self):
        super().__init__()
        self.k = 1.0
    def forward(self, x):
        if self.training:
            assert not torch.jit.is_scripting(), "this codepath is not supported"
            return x * self.k if self.k != "test" else x   # error in torchscript due to type mistmatch
        else:
            return x + 1

a = A()
a = torch.jit.script(a)
```

The former case can be used as a workaround but it's not an ideal solution, because it does not throw errors when user tries to use the scripted model in training mode. Bypassing code sections using `assert not torch.jit.is_scripting()` seems like the simplest syntax: it's only one line of code and does not introduce extra indentation. This is similar to `@torch.jit.unused`: it skips the code with one extra line and throws an error if the skipped code is called.
(This issue is not about union type specifically, since it applies to other python features missing in torchscript.)

## Alternatives
```python
class A(nn.Module):
    def __init__(self):
        super().__init__()
        self.k = 1.0
    def forward(self, x):
        if self.training:
            if not torch.jit.is_scripting():
                return x * self.k if self.k != "test" else x
            else:
                assert False, "this codepath is not supported"
        else:
            return x + 1
```
this works now but it's more ugly.

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by ppwwyyxx**
Summary: fix https://github.com/pytorch/pytorch/issues/44864

Test Plan: buck test mode/dev-nosan //caffe2/test:jit -- 'test_assert_is_script'

Differential Revision: D23755094

### CODE REVIEW

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 1]
```diff
@@ -1082,6 +1082,9 @@ struct to_ir {
           throw ErrorReport(stmt)
               << "Unrecognized statement kind " << kindToString(stmt.kind());
       }
+      // Already found an exit statement in this block. Do not parse further
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1085]: nit: we've already parsed the remaining block of code, here we're not emitting further because it's not reachable. maybe add a small reference that the remaining statements aren't reachable

**Review by eellison — APPROVED** [review]
LGTM! Nice change!
